cpp/src/arrow/engine/substrait/extension_set.cc

```diff
@@ -1043,6 +1049,10 @@ struct DefaultExtensionIdRegistry : ExtensionIdRegistryImpl {
              TypeName{uint32(), "u32"},
              TypeName{uint64(), "u64"},
              TypeName{float16(), "fp16"},
+             TypeName{large_utf8(), "large_string"},
+             TypeName{large_binary(), "large_binary"},
+             TypeName{date64(), "date_millis"},
+             TypeName{time64(TimeUnit::NANO), "time_nanos"},
```

Comment:
It is not obvious to me that we can't add time32:
```suggestion
             TypeName{time32(TimeUnit::SECONDS), "time_seconds"},
```

Reply:
During encoding we need to figure out the type's "name".  We do this by looking into the registry.  The registry, for types, holds:

```
std::unordered_map<const DataType*, int, TypePtrHashEq, TypePtrHashEq> type_to_index_;
```
As a result, we can only register a single UDT name per `DataType`.  We can get away with it in `time64` because `time64(TimeUnit::MICROS)` doesn't go down the UDT path and doesn't use the above map.

With `time32` we could only support one of `TimeUnit::SECOND` or `TimeUnit::MILLISECONDS`.  I suppose I could have just picked one but I went with the "let's just solve this when we tackle parameterized UDTs" route.

Reply:
Well, I will eat my words.  Your suggestion worked.